File: ambeth/Ambeth.Merge/ambeth/merge/model/CreateOrUpdateContainerBuild.cs
using De.Osthus.Ambeth.Collections;
using De.Osthus.Ambeth.Merge.Model;
using De.Osthus.Ambeth.Merge.Transfer;
using De.Osthus.Ambeth.Util;
using System;
using System.Text;

namespace De.Osthus.Ambeth.Merge.Model
{
    public class CreateOrUpdateContainerBuild : AbstractChangeContainer, ICreateOrUpdateContainer
    {
        protected IPrimitiveUpdateItem[] fullPUIs;

        protected readonly HashMap<String, int?> relationNameToIndexMap;

        protected readonly HashMap<String, int?> primitiveNameToIndexMap;

        protected IRelationUpdateItem[] fullRUIs;

        protected int ruiCount, puiCount;

        protected bool isCreate;

        protected readonly ICUDResultHelper cudResultHelper;

        public CreateOrUpdateContainerBuild(bool isCreate, HashMap<String, int?> relationNameToIndexMap, HashMap<String, int?> primitiveNameToIndexMap,
            ICUDResultHelper cudResultHelper)
        {
            this.isCreate = isCreate;
            this.relationNameToIndexMap = relationNameToIndexMap;
            this.primitiveNameToIndexMap = primitiveNameToIndexMap;
            this.cudResultHelper = cudResultHelper;
        }

        public bool IsCreate()
        {
            return isCreate;
        }

        public bool IsUpdate()
        {
            return !IsCreate();
        }

        public IPrimitiveUpdateItem[] GetFullPUIs()
        {
            return fullPUIs;
        }

        public IRelationUpdateItem[] GetFullRUIs()
        {
            return fullRUIs;
        }

        public void AddPrimitive(IPrimitiveUpdateItem pui)
        {
            IPrimitiveUpdateItem[] fullPUIs = this.fullPUIs;
            if (fullPUIs == null)
            {
                fullPUIs = new IPrimitiveUpdateItem[primitiveNameToIndexMap.Count];
                this.fullPUIs = fullPUIs;
            }
            int? indexR = primitiveNameToIndexMap.Get(pui.MemberName);
            if (!indexR.HasValue)
            {
                throw new Exception("No primitive member " + Reference.RealType.FullName + "." + pui.MemberName + " defined");
            }
            int index = indexR.Value;
            if (fullPUIs[index] == null)
            {
                puiCount++;
            }
            fullPUIs[index] = pui;
        }

        public void AddRelation(IRelationUpdateItem rui)
        {
            IRelationUpdateItem[] fullRUIs = this.fullRUIs;
            if (fullRUIs == null)
            {
                fullRUIs = new IRelationUpdateItem[relationNameToIndexMap.Count];
                this.fullRUIs = fullRUIs;
            }
            int? indexR = relationNameToIndexMap.Get(rui.MemberName);
            if (!indexR.HasValue)
            {
                throw new Exception("No relation member " + Reference.RealType.FullName + "." + rui.MemberName + " defined");
            }
            int index = indexR.Value;
            if (fullRUIs[index] == null)
            {
                ruiCount++;
            }
            fullRUIs[index] = rui;
        }

        public PrimitiveUpdateItem FindPrimitive(String memberName)
        {
            if (fullPUIs == null)
            {
                return null;
            }
            int? indexR = primitiveNameToIndexMap.Get(memberName);
            if (!indexR.HasValue)
            {
                throw new Exception("No primitive member " + Reference.RealType.FullName + "." + memberName + " defined");
            }
            return (PrimitiveUpdateItem)fullPUIs[indexR.Value];
        }

        public RelationUpdateItemBuild FindRelation(String memberName)
        {
            if (fullRUIs == null)
            {
                return null;
            }
            int? indexR = relationNameToIndexMap.Get(memberName);
            if (!indexR.HasValue)
            {
                throw new Exception("No relation member " + Reference.RealType.FullName + "." + memberName + " defined");
            }
            return (RelationUpdateItemBuild)fullRUIs[indexR.Value];
        }

        public PrimitiveUpdateItem EnsurePrimitive(String memberName)
        {
            PrimitiveUpdateItem pui = FindPrimitive(memberName);
            if (pui != null)
            {
                return pui;
            }
            pui = new PrimitiveUpdateItem();
            pui.MemberName = memberName;
            AddPrimitive(pui);
            return pui;
        }

        public RelationUpdateItemBuild EsureRelation(String memberName)
        {
            RelationUpdateItemBuild rui = FindRelation(memberName);
            if (rui != null)
            {
                return rui;
            }
            rui = new RelationUpdateItemBuild(memberName);
            AddRelation(rui);
            return rui;
        }

        public int GetPuiCount()
        {
            return puiCount;
        }

        public int GetRuiCount()
        {
            return ruiCount;
        }

        public override void ToString(StringBuilder sb)
	    {
		    if (IsCreate())
		    {
			    sb.Append(typeof(CreateContainer).Name).Append(": ");
		    }
		    else if (IsUpdate())
		    {
			    sb.Append(typeof(UpdateContainer).Name).Append(": ");
		    }
		    else
		    {
			    base.ToString(sb);
			    return;
		    }
		    StringBuilderUtil.AppendPrintable(sb, Reference);
	    }

	    public ICreateOrUpdateContainer Build()
	    {
            if (IsCreate())
		    {
			    CreateContainer cc = new CreateContainer();
			    cc.Reference = Reference;
			    cc.Primitives = cudResultHelper.CompactPUIs(GetFullPUIs(), GetPuiCount());
			    cc.Relations = cudResultHelper.CompactRUIs(GetFullRUIs(), GetRuiCount());
			    return cc;
		    }
            if (IsUpdate())
		    {
			    UpdateContainer uc = new UpdateContainer();
			    uc.Reference = Reference;
			    uc.Primitives = cudResultHelper.CompactPUIs(GetFullPUIs(), GetPuiCount());
			    uc.Relations = cudResultHelper.CompactRUIs(GetFullRUIs(), GetRuiCount());
			    return uc;
		    }
		    throw new Exception("Must never happen");
	    }
    }
}

File: jambeth/jambeth-merge/src/main/java/de/osthus/ambeth/merge/model/CreateOrUpdateContainerBuild.java
package de.osthus.ambeth.merge.model;

import de.osthus.ambeth.collections.HashMap;
import de.osthus.ambeth.merge.ICUDResultHelper;
import de.osthus.ambeth.merge.transfer.AbstractChangeContainer;
import de.osthus.ambeth.merge.transfer.CreateContainer;
import de.osthus.ambeth.merge.transfer.PrimitiveUpdateItem;
import de.osthus.ambeth.merge.transfer.UpdateContainer;
import de.osthus.ambeth.proxy.IEntityMetaDataHolder;
import de.osthus.ambeth.util.StringBuilderUtil;

public class CreateOrUpdateContainerBuild extends AbstractChangeContainer implements ICreateOrUpdateContainer, IEntityMetaDataHolder
{
	protected IPrimitiveUpdateItem[] fullPUIs;

	protected final HashMap<String, Integer> relationNameToIndexMap;

	protected final HashMap<String, Integer> primitiveNameToIndexMap;

	protected IRelationUpdateItem[] fullRUIs;

	protected final IEntityMetaData metaData;

	protected int ruiCount, puiCount;

	protected boolean isCreate;

	protected final ICUDResultHelper cudResultHelper;

	public CreateOrUpdateContainerBuild(IEntityMetaData metaData, boolean isCreate, HashMap<String, Integer> relationNameToIndexMap,
			HashMap<String, Integer> primitiveNameToIndexMap, ICUDResultHelper cudResultHelper)
	{
		this.metaData = metaData;
		this.isCreate = isCreate;
		this.relationNameToIndexMap = relationNameToIndexMap;
		this.primitiveNameToIndexMap = primitiveNameToIndexMap;
		this.cudResultHelper = cudResultHelper;
	}

	@Override
	public IEntityMetaData get__EntityMetaData()
	{
		return metaData;
	}

	public boolean isCreate()
	{
		return isCreate;
	}

	public boolean isUpdate()
	{
		return !isCreate();
	}

	@Override
	public IPrimitiveUpdateItem[] getFullPUIs()
	{
		return fullPUIs;
	}

	@Override
	public IRelationUpdateItem[] getFullRUIs()
	{
		return fullRUIs;
	}

	public void addPrimitive(IPrimitiveUpdateItem pui)
	{
		IPrimitiveUpdateItem[] fullPUIs = this.fullPUIs;
		if (fullPUIs == null)
		{
			fullPUIs = new IPrimitiveUpdateItem[primitiveNameToIndexMap.size()];
			this.fullPUIs = fullPUIs;
		}
		Integer indexR = primitiveNameToIndexMap.get(pui.getMemberName());
		if (indexR == null)
		{
			throw new IllegalStateException("No primitive member " + getReference().getRealType().getName() + "." + pui.getMemberName() + " defined");
		}
		int index = indexR.intValue();
		if (fullPUIs[index] == null)
		{
			puiCount++;
		}
		fullPUIs[index] = pui;
	}

	public void addRelation(IRelationUpdateItem rui)
	{
		IRelationUpdateItem[] fullRUIs = this.fullRUIs;
		if (fullRUIs == null)
		{
			fullRUIs = new IRelationUpdateItem[relationNameToIndexMap.size()];
			this.fullRUIs = fullRUIs;
		}
		Integer indexR = relationNameToIndexMap.get(rui.getMemberName());
		if (indexR == null)
		{
			throw new IllegalStateException("No relation member " + getReference().getRealType().getName() + "." + rui.getMemberName() + " defined");
		}
		int index = indexR.intValue();
		if (fullRUIs[index] == null)
		{
			ruiCount++;
		}
		fullRUIs[index] = rui;
	}

	public PrimitiveUpdateItem findPrimitive(String memberName)
	{
		if (fullPUIs == null)
		{
			return null;
		}
		Integer indexR = primitiveNameToIndexMap.get(memberName);
		if (indexR == null)
		{
			throw new IllegalStateException("No primitive member " + getReference().getRealType().getName() + "." + memberName + " defined");
		}
		return (PrimitiveUpdateItem) fullPUIs[indexR.intValue()];
	}

	public RelationUpdateItemBuild findRelation(String memberName)
	{
		if (fullRUIs == null)
		{
			return null;
		}
		Integer indexR = relationNameToIndexMap.get(memberName);
		if (indexR == null)
		{
			throw new IllegalStateException("No relational member " + getReference().getRealType().getName() + "." + memberName + " defined");
		}
		return (RelationUpdateItemBuild) fullRUIs[indexR.intValue()];
	}

	public PrimitiveUpdateItem ensurePrimitive(String memberName)
	{
		PrimitiveUpdateItem pui = findPrimitive(memberName);
		if (pui != null)
		{
			return pui;
		}
		pui = new PrimitiveUpdateItem();
		pui.setMemberName(memberName);
		addPrimitive(pui);
		return pui;
	}

	public RelationUpdateItemBuild ensureRelation(String memberName)
	{
		RelationUpdateItemBuild rui = findRelation(memberName);
		if (rui != null)
		{
			return rui;
		}
		rui = new RelationUpdateItemBuild(memberName);
		addRelation(rui);
		return rui;
	}

	public int getPuiCount()
	{
		return puiCount;
	}

	public int getRuiCount()
	{
		return ruiCount;
	}

	@Override
	public void toString(StringBuilder sb)
	{
		if (isCreate())
		{
			sb.append(CreateContainer.class.getSimpleName()).append(": ");
		}
		else if (isUpdate())
		{
			sb.append(UpdateContainer.class.getSimpleName()).append(": ");
		}
		else
		{
			super.toString(sb);
			return;
		}
		StringBuilderUtil.appendPrintable(sb, reference);
	}

	public ICreateOrUpdateContainer build()
	{
		if (isCreate())
		{
			CreateContainer cc = new CreateContainer();
			cc.setReference(getReference());
			cc.setPrimitives(cudResultHelper.compactPUIs(getFullPUIs(), getPuiCount()));
			cc.setRelations(cudResultHelper.compactRUIs(getFullRUIs(), getRuiCount()));
			return cc;
		}
		if (isUpdate())
		{
			UpdateContainer uc = new UpdateContainer();
			uc.setReference(getReference());
			uc.setPrimitives(cudResultHelper.compactPUIs(getFullPUIs(), getPuiCount()));
			uc.setRelations(cudResultHelper.compactRUIs(getFullRUIs(), getRuiCount()));
			return uc;
		}
		throw new IllegalStateException("Must never happen");
	}
}
